substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,667 @@
-		(ffi_value, data).into()
+	fn into_ffi_value(value: &mut Self::Inner) -> (Self::FFIType, Self::Destructor) {
+		(pack_ptr_and_len(value.as_ptr() as u32, value.len() as u32), ())
 	}
+}
 
-	fn from_ffi_value(arg: Self::FFIType) -> T {
-		let (ptr, len) = unpack_ptr_and_len(arg);
-		let len = len as usize;
+/// Pass a pointer into the host and write to it after the host call ends.
+///
+/// This casts a given type into `&mut [u8]` using `AsMut<[u8]>` and passes a pointer to
+/// that byte slice into the host. The host *doesn't* read from this and instead creates
+/// a default instance of type `T` and passes it as a `&mut T` into the host function
+/// implementation. After the host function finishes this value is then cast into a `&[u8]` using
+/// `AsRef<[u8]>` and written back into the guest memory.
+///
+/// Raw FFI type: `u32` (a pointer)
+pub struct PassPointerAndWrite<T, const N: usize>(PhantomData<(T, [u8; N])>);
 
-		let encoded = if len == 0 {
-			bytes::Bytes::new()
-		} else {
-			bytes::Bytes::from(unsafe { Vec::from_raw_parts(ptr as *mut u8, len, len) })
-		};
+impl<T, const N: usize> RIType for PassPointerAndWrite<T, N> {
+	type FFIType = u32;
+	type Inner = T;
+}
 
-		codec::decode_from_bytes(encoded).expect("Host to wasm values are encoded correctly; qed")
+#[cfg(not(substrate_runtime))]
+impl<'a, T, const N: usize> FromFFIValue<'a> for PassPointerAndWrite<&'a mut T, N>
+where
+	T: Default + AsRef<[u8]>,
+{
+	type Owned = T;
+
+	fn from_ffi_value(
+		_context: &mut dyn FunctionContext,
+		_arg: Self::FFIType,
+	) -> Result<Self::Owned> {
+		Ok(T::default())
+	}
+
+	fn take_from_owned(owned: &'a mut Self::Owned) -> Self::Inner {
+		&mut *owned
+	}
+
+	fn write_back_into_runtime(
+		value: Self::Owned,
+		context: &mut dyn FunctionContext,
+		arg: Self::FFIType,
+	) -> Result<()> {
+		let value = value.as_ref();
+		assert_eq!(value.len(), N);
+		context.write_memory(Pointer::new(arg), value)
 	}
 }
 
-/// The type is passed as `u64`.
+#[cfg(substrate_runtime)]
+impl<'a, T, const N: usize> IntoFFIValue for PassPointerAndWrite<&'a mut T, N>
+where
+	T: AsMut<[u8]>,
+{
+	type Destructor = ();
+
+	fn into_ffi_value(value: &mut Self::Inner) -> (Self::FFIType, Self::Destructor) {
+		let value = value.as_mut();
+		assert_eq!(value.len(), N);
+		(value.as_ptr() as u32, ())
+	}
+}
+
+/// Pass a `T` into the host using the SCALE codec.
 ///
-/// The `u64` value is build by `length 32bit << 32 | pointer 32bit`
+/// This encodes a `T` into a `Vec<u8>` using the SCALE codec and then
+/// passes a pointer to that byte blob and its length to the host,
+/// which then reads it and decodes back into `T`.
 ///
-/// `Self` is encoded and the length and the pointer are taken from the encoded vector.
-impl<T: codec::Codec> RIType for Codec<T> {
+/// Raw FFI type: `u64` (a fat pointer; upper 32 bits is the size, lower 32 bits is the pointer)
+pub struct PassByCodec<T>(PhantomData<T>);
```

Comment:
Nit: Maybe `PassFatPointerEncoded`?

Reply:
I renamed `PassSliceRefByCodec` -> `PassFatPointerAndDecodeSlice` and `PassByCodec` -> `PassFatPointerAndDecode`.

These are might be a little too long, but at least they're consistent with the others.